tracing-subscriber/src/filter/field.rs

```diff
@@ -0,0 +1,215 @@
+use matchers::Pattern;
+use std::{
+    error::Error,
+    fmt,
+    str::FromStr,
+    sync::atomic::{AtomicBool, Ordering},
+};
+
+use super::{FieldMap, LevelFilter};
+use tracing_core::field::{Field, Visit};
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct Match {
+    pub(crate) name: String, // TODO: allow match patterns for names?
+    pub(crate) value: Option<ValueMatch>,
+}
+
+#[derive(Debug, Eq, PartialEq)]
+pub struct CallsiteMatch {
+    pub(crate) fields: FieldMap<ValueMatch>,
+    pub(crate) level: LevelFilter,
+}
+
+#[derive(Debug)]
+pub struct SpanMatch {
+    fields: FieldMap<(ValueMatch, AtomicBool)>,
+    level: LevelFilter,
+    has_matched: AtomicBool,
+}
+
+pub struct MatchVisitor<'a> {
+    inner: &'a SpanMatch,
+}
+
+#[derive(Debug, Clone)]
+pub(crate) enum ValueMatch {
+    Bool(bool),
+    U64(u64),
+    I64(i64),
+    Pat(Pattern),
+}
+
+#[derive(Clone, Debug)]
+pub struct BadName {
+    name: String,
+}
+
+// === impl Match ===
+
+impl FromStr for Match {
+    type Err = Box<dyn Error + Send + Sync>;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.split('=');
+        let name = parts
+            .next()
+            .ok_or_else(|| BadName {
+                name: "".to_string(),
+            })?
+            // TODO: validate field name
+            .to_string();
+        let value = parts.next().map(ValueMatch::from_str).transpose()?;
+        Ok(Match { name, value })
+    }
+}
+
+impl Match {
+    pub fn has_value(&self) -> bool {
+        self.value.is_some()
+    }
+
+    // TODO: reference count these strings?
+    pub fn name(&self) -> String {
+        self.name.clone()
+    }
+}
+
+// === impl ValueMatch ===
+
+impl FromStr for ValueMatch {
+    type Err = matchers::Error;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        s.parse::<bool>()
+            .map(ValueMatch::Bool)
+            .or_else(|_| s.parse::<u64>().map(ValueMatch::U64))
+            .or_else(|_| s.parse::<i64>().map(ValueMatch::I64))
+            .or_else(|_| s.parse::<Pattern>().map(ValueMatch::Pat))
+    }
+}
+
+impl PartialEq<Self> for ValueMatch {
+    fn eq(&self, other: &Self) -> bool {
+        match (self, other) {
+            (ValueMatch::Bool(a), ValueMatch::Bool(b)) => a == b,
+            (ValueMatch::I64(a), ValueMatch::I64(b)) => a == b,
+            (ValueMatch::U64(a), ValueMatch::U64(b)) => a == b,
+            // TODO: T_T
```

Comment:
oh, we should be testing equality of patterns for this to truly be correct, but i never wrote it. i'll fix that,

Reply:
ah, right: the issue is that we can't easily implement equality for this type, since the `regex_automata::DenseDFA` that represents the pattern internally doesn't impl eq. in practice, we won't use this `Eq` impl ever, but it would be nice if it was correct. I'll update the comment to reflect that.